### PR TITLE
[pytorch] fix ParallelNative.h/cpp build

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26702 [pytorch][perf] add mobile friendly at:parallel_for backend
* **#26746 [pytorch] fix ParallelNative.h/cpp build**
* #26783 [pytorch][mobile] remove backward functions from jit-op-registry for mobile build

Summary:
PR #25439 introduced type() to IValue, which enforced ivalue::Future to
be constructed with type. This broke ParallelNative.h and a few other
places where Future was constructed without type and simply used as a barrier:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelNative.h#L63

Use NoneType::get() to initialize these Futures.

Test Plan:
- builds

Differential Revision: [D17556002](https://our.internmc.facebook.com/intern/diff/D17556002)

### CODE REVIEW

File: aten/src/ATen/core/ivalue_inl.h [hunk 1]
```diff
@@ -191,6 +191,7 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
   }
 
  public:
+  Future() {}
```
Comment (smessmer) [inline aten/src/ATen/core/ivalue_inl.h]: nit: `Future() = default;` is better

**Review by smessmer — COMMENTED** [review]
Are we sure we want to be able to construct Futures without type information? Afaik, we want to have types everywhere. cc @zdevito

**Review by suo — CHANGES_REQUESTED** [review]
I don't think this makes a ton of sense. TupleType takes a null type because the type can be lazily inferred by inspecting the types of the contained elements, not because it actually doesn't have a type.

I would prefer not to default construct `Future` as there is not an obvious default. Can we change ParallelNative and friends to use `Future<None>`?

**Comment by ilia-cher** [comment]
@suo I wonder would it make sense to make it None by default (with default ctor)

File: aten/src/ATen/ParallelNative.h [hunk 2]
```diff
@@ -60,7 +60,10 @@ inline void parallel_for(
       internal::_unset_thread_num();
     };
 
-    std::vector<c10::ivalue::Future> futures(num_tasks);
+    std::vector<std::shared_ptr<c10::ivalue::Future>> futures(num_tasks);
+    for (size_t i = 0; i < num_tasks; ++i) {
+      futures[i] = std::make_shared<c10::ivalue::Future>(NoneType::get());
+    }
```
Comment (ljk53) [inline aten/src/ATen/ParallelNative.h]: @ilia-cher let me know if I should move this to a common util function, maybe in Parallel.h? I can reuse it in ParallelNativeMobile.h as well.

**Comment by zdevito** [comment]
This and https://github.com/pytorch/pytorch/pull/26739 are duplicates, fyi.

**Pull request closed without merging**